iceberg-rust/src/object_store/mod.rs

```diff
@@ -125,6 +151,9 @@ impl ObjectStoreBuilder {
     /// Set config value for builder
     pub fn with_config(self, key: ConfigKey, value: impl Into<String>) -> Self {
         match (self, key) {
+            (ObjectStoreBuilder::Azure(azure), ConfigKey::Azure(key)) => {
+                ObjectStoreBuilder::Azure(Box::new(azure.with_config(key, value)))
+            }
             (ObjectStoreBuilder::S3(aws), ConfigKey::AWS(key)) => {
                 ObjectStoreBuilder::S3(Box::new(aws.with_config(key, value)))
             }
```

Comment:
I realized that it's not good to parse the config string into a `ConfigKey` before passing it to the builder, because there is no knowledge about which provider(s3,gcs,azure) to parse it for. This is an issue if the providers have matching config keys. That's why one of the tests is failing.
Instead we need to pass the string to `with_config()` and then parse the `ConfigKey` inside of it when we have the appropriate information.
I'll create a PR to fix this.

Reply:
In our system, we use the `ObjectStoreScheme::parse()` with the URL to define which builder to use.

Reply:
Both ways. We have some fixed options that we pass to the builder in a structured way, but the user can set more and those we add with `with_config()`.